orangecontrib/text/ontology.py

```diff
@@ -0,0 +1,381 @@
+    crossover_probability: float = 0.8,
+    mutation_probability: float = 0.1,
+    callback: Callable = dummy_callback
+) -> Tuple[np.array, int]:
+    seq_len = sims.shape[0] - 2
+    rng = np.random.default_rng(0)
+    generation = rng.integers(seq_len, size=(population_size, seq_len))
+    fitness_function = FitnessFunction(words, sims).fitness
+    ticks = iter(np.linspace(0.2, 1.0, num_generations))
+
+    for _ in range(num_generations):
+        callback(next(ticks))
+        fitness = [
+            fitness_function(Tree.from_prufer_sequence(generation[i, :], words))[0]
+            for i in range(population_size)
+        ]
+        new_generation = np.zeros(generation.shape, dtype=np.int32)
+        for i in range(population_size):
+            if rng.uniform() > crossover_probability:
+                new_generation[i, :] = generation[rng.integers(seq_len), :]
+            else:
+                u, v = rng.integers(seq_len, size=2)
+                point = rng.integers(seq_len // 2)
+                new_generation[i, :point] = generation[u, :point]
+                new_generation[i, point:point+seq_len//2] = generation[v, point:point+seq_len//2]
+                new_generation[i, point+seq_len//2:] = generation[u, point+seq_len//2:]
+            mutation_idx = np.where(rng.uniform(size=seq_len) < mutation_probability)[0]
+            new_generation[i, mutation_idx] = rng.integers(seq_len, size=len(mutation_idx))
+        generation = new_generation
+
+    # select the best solution
+    scores, roots = list(), list()
+    for i in range(population_size):
+        score, root = fitness_function(Tree.from_prufer_sequence(generation[i, :], words))
+        scores.append(score)
+        roots.append(root)
+    best = np.argmax(scores)
+    return generation[best, :], roots[best]
+
+
+def cos_sim(x: np.array, y: np.array) -> float:
+    dot = np.dot(x, y)
+    return 0 if np.allclose(dot, 0) else dot / (np.linalg.norm(x) * np. linalg.norm(y))
+
+
+class EmbeddingStorage:
+
+    def __init__(self):
+        self.cache_dir = os.path.join(cache_dir(), 'ontology')
+        if not os.path.isdir(self.cache_dir):
+            os.makedirs(self.cache_dir)
+        self.similarities = dict()
+        try:
+            with open(os.path.join(self.cache_dir, 'sims.pkl'), 'rb') as file:
+                self.similarities = pickle.load(file)
+        except Exception as _:
+            self.similarities = dict()
+        self.embeddings = dict()
+
+    def save_similarities(self):
+        with open(os.path.join(self.cache_dir, 'sims.pkl'), 'wb') as file:
+            pickle.dump(self.similarities, file)
+
+    def get_embedding(self, word: str) -> Optional[np.array]:
+        if word in self.embeddings:
+            return self.embeddings[word]
+        try:
+            emb = np.load(os.path.join(self.cache_dir, f'{word}.npy'))
+            self.embeddings[word] = emb
+            return emb
+        except Exception as _:
```

Comment:
I think catching the IOError error is enough here. Also you can skip `as` part if you do not need an exception
```suggestion
        except IOError:
```